src/pudl/transform/eia861.py

```diff
@@ -193,79 +614,708 @@ def service_territory(raw_dfs, tfr_dfs):
+
+    Returns:
+        dict: a dictionary of transformed dataframes. This function
+        both compiles the association table, and finishes the
+        normalization of the balancing authority table. It may be that
+        once the harvesting process incorporates the EIA 861, some or
+        all of this functionality should be pulled into the phase-2
+        transform functions.
+
+    """
+    # The dataframes from which to compile BA-Util-State associations
+    other_dfs = [
+        tfr_dfs["sales_eia861"],
+        tfr_dfs["demand_response_eia861"],
+        tfr_dfs["advanced_metering_infrastructure_eia861"],
+        tfr_dfs["dynamic_pricing_eia861"],
+    ]
+
+    logger.info("Building an EIA 861 BA-Util-State association table.")
+
+    old_date_ba_util = (
+        tfr_dfs["balancing_authority_eia861"]
+        .query("report_date<='2012-12-31'")
+        .loc[:, [
+            "report_date",
+            "balancing_authority_id_eia",
+            "utility_id_eia",
+        ]]
+        .dropna()
+        .drop_duplicates()
+    )
+
+    old_date_util_state = pd.DataFrame()
+    for df in other_dfs:
+        tmp_df = (
+            df.query("report_date<='2012-12-31'")
+            .loc[:, [
+                "report_date",
+                "utility_id_eia",
+                "state",
+            ]]
+            .dropna()
+            .drop_duplicates()
+        )
+        old_date_util_state = (
+            pd.concat([old_date_util_state, tmp_df])
+            .drop_duplicates()
+        )
+    old_date_ba_util_state = pd.merge(
+        old_date_ba_util, old_date_util_state, how="outer")
+
+    new_ba_code_id = (
+        tfr_dfs["balancing_authority_eia861"]
+        .query("report_date>='2013-01-01'")
+        .loc[:, [
+            "report_date",
+            "balancing_authority_code_eia",
+            "balancing_authority_id_eia",
+        ]]
+        .dropna()
+        .drop_duplicates()
+    )
+    new_date_ba_code_util_state = pd.DataFrame()
+    for df in other_dfs:
+        tmp_df = (
+            df.query("report_date>='2013-01-01'")
+            .loc[:, [
+                "report_date",
+                "balancing_authority_code_eia",
+                "utility_id_eia",
+                "state",
+            ]]
+            .dropna()
+            .drop_duplicates()
+        )
+        new_date_ba_code_util_state = (
+            pd.concat([new_date_ba_code_util_state, tmp_df])
+            .drop_duplicates()
+        )
+    new_date_ba_util_state = (
+        new_date_ba_code_util_state
+        .merge(new_ba_code_id, how="outer")
+        .drop("balancing_authority_code_eia", axis="columns")
+        .drop_duplicates()
+    )
+    tfr_dfs["balancing_authority_assn_eia861"] = (
+        pd.concat([old_date_ba_util_state, new_date_ba_util_state])
+        .dropna(subset=["balancing_authority_id_eia", ])
+        .astype({"utility_id_eia": pd.Int64Dtype()})
+    )
+
+    # Finish the normalization of the BA table:
```

Comment:
Does this section below here need to be in this function? It looks like it could be separated, which would be nice bc it feels pretty different.

Reply:
Maybe this wasn't clear but, this is here because the transformation of the `balancing_authority_eia861` table can't be finished until these associations have been extracted from it -- the structure that it has up until this point is different.  But we can't generate the associations table until all of the other data tables have been transformed, so we can see how various BAs/Utils/States show up with each other. So the `balancing_authority_assn_eia861` function ends up both creating the association table, and finishing the transformation of the original `balancing_authority_eia861` table. Are you saying that you would prefer a second transform function that is called after this one, which completes the transformation of the ba table? Or that you want this little bit put in a separate function which is still called from within here?

Reply:
the requirement for the ordering of these things was not clear to me. and I think it would be easier to understand if there was a second transform function run after this one. Perhaps you could check if the association table exists before generating the second step so you don't accidentally run step 2 first.